### PR TITLE
Use MapVector to fix lld thinLTO "nondeterminism" issue.

### PR DESCRIPTION
**Pull request by llvmssh**
When the ModuleSymbolTable is generated, the binary consistency problem occurs due to the
data structure for collecting asm symbols was ordered by memory pointers.
Use MapVector to fix lld thinLTO "nondeterminism" issue.

### CODE REVIEW

**Comment by jh7370** [comment]
I don't know who filed this PR, but the author is listed as "None". I think whoever did this has messed up their commit email/author details?

**Comment by llvmssh** [comment]
> I don't know who filed this PR, but the author is listed as "None". I think whoever did this has messed up their commit email/author details?

Thank you for your comment. I should have modified the information.

**Comment by llvmssh** [comment]
> Thanks for the update. Is there a GitHub issue associated with the problem you're trying to fix? If so, please include "Fixes #####" in your PR description.
> 
> Just so you know, I'm not familiar with the area, so somebody who is will need to perform the review. Take a look through past contributors to this file and try pinging them on here/adding them as reviewers.

Thanks, I have a problem, it seems that I can't add reviewers.

**Comment by jh7370** [comment]
> > Thanks for the update. Is there a GitHub issue associated with the problem you're trying to fix? If so, please include "Fixes #####" in your PR description.
> > Just so you know, I'm not familiar with the area, so somebody who is will need to perform the review. Take a look through past contributors to this file and try pinging them on here/adding them as reviewers.
> 
> Thanks, I have a problem, it seems that I can't add reviewers.

From the automated greeting post above:

> If you wish to, you can add reviewers by using the "Reviewers" section on this page.
>
> If this is not working for you, it is probably because you do not have write permissions for the repository. In which case you can instead tag reviewers by name in a comment by using @ followed by their GitHub username.

**Comment by llvmssh** [comment]
@vitalybuka Hi，I see you have modified this file before, I have made a little modification, can you help review it?

**Comment by llvmssh** [comment]
@pcc，@vitalybuka Hi, can you help review my code?  thank you.
Fix the issue:
During the generation of ModuleSymbolTable, the unordered DenseMap data structure is traversed during the collection of Asm symbols. As a result, the sequence of symbols in the symbol table is inconsistent after the ThinLTO lld link. As a result, binary random inconsistency occurs. Made a simple data structure modification.

**Comment by llvmssh** [comment]
> > > Thanks for the update. Is there a GitHub issue associated with the problem you're trying to fix? If so, please include "Fixes #####" in your PR description.
> > > Just so you know, I'm not familiar with the area, so somebody who is will need to perform the review. Take a look through past contributors to this file and try pinging them on here/adding them as reviewers.
> > 
> > 
> > Thanks, I have a problem, it seems that I can't add reviewers.
> 
> From the automated greeting post above:
> 
> > If you wish to, you can add reviewers by using the "Reviewers" section on this page.
> > If this is not working for you, it is probably because you do not have write permissions for the repository. In which case you can instead tag reviewers by name in a comment by using @ followed by their GitHub username.

Can you add a few reviewers for me?

**Comment by vitalybuka** [comment]
What is "bep issue"?

**Comment by llvmssh** [comment]
> What is "bep issue"?

Thanks for your reply. When a compiler of the same version builds the same project twice, the generated binary files are different, which is reflected in the inconsistent sequence in the symbol table.  For example:
< 295: 000000000000 0 FUNC  GLOBAL DEFAULT UND __lxstat@GLIBC_2.X 
>295: 000000000000  0 FUNC  GLOBAL DEFAULT UND __libc_start_main@GLIBC_2.X

**Comment by vitalybuka** [comment]
> Thanks for your reply. When a compiler of the same version builds the same project twice, the generated binary files are different, which is reflected in the inconsistent sequence in the symbol table. For example: 295: 000000000000 0 FUNC GLOBAL DEFAULT UND [__lxstat@GLIBC_2.X](mailto:__lxstat@GLIBC_2.X) 295: 000000000000 0 FUNC GLOBAL DEFAULT UND [__libc_start_main@GLIBC_2.X](mailto:__libc_start_main@GLIBC_2.X)

Sorry, out of curiosity are there definition of "bep"? 

Regardless of definition, can we please just call it "nondeterminism" in title and description? I guess that's how we usually call it in LLVM.

File: llvm/lib/Object/RecordStreamer.h [hunk 1]
```diff
@@ -28,11 +29,11 @@ class RecordStreamer : public MCStreamer {
 
 private:
   const Module &M;
-  StringMap<State> Symbols;
+  MapVector<StringRef, State> Symbols;
```
Comment (vitalybuka) [inline llvm/lib/Object/RecordStreamer.h:32]: I would not expect nondeterminism from Symbols
Order in hash table should be determined character in StringRefs and order they inserted.
Should be the same all the time?

Comment (llvmssh) [reply llvm/lib/Object/RecordStreamer.h:32]: Symbols generation order comes from traversing SymverAliasMap in flushSymverDirectives func, which is an unordered map data structure, and I understand that traversing creates nondeterminism problems.

File: llvm/lib/Object/RecordStreamer.h [hunk 2]
```diff
@@ -28,11 +29,11 @@ class RecordStreamer : public MCStreamer {
...
   // Map of aliases created by .symver directives, saved so we can update
   // their symbol binding after parsing complete. This maps from each
   // aliasee to its list of aliases.
-  DenseMap<const MCSymbol *, std::vector<StringRef>> SymverAliasMap;
+  MapVector<const MCSymbol *, std::vector<StringRef>> SymverAliasMap;
```
Comment (vitalybuka) [inline llvm/lib/Object/RecordStreamer.h:36]: This one I see that it was non deterministic `DenseMap<const MCSymbol *`

Comment (llvmssh) [reply llvm/lib/Object/RecordStreamer.h:36]: yes, this may cause nondeterminism problem.

File: llvm/lib/Object/RecordStreamer.h (hunk 1 above)
Comment (vitalybuka) [reply llvm/lib/Object/RecordStreamer.h:32]: unordered does not mean nondeterminism.

So I believe change in this line is unnecessary.

**Comment by vitalybuka** [comment]
@teresajohnson who added the loop in https://reviews.llvm.org/D30485

File: llvm/lib/Object/RecordStreamer.h (hunk 1 above)
Comment (vitalybuka) [reply llvm/lib/Object/RecordStreamer.h:32]: To clarify, order is determined by:
1. order and values passed into markDefined/markDefined/markUsed invoked
2. hash table implementation

So for the same lld binary those are consistent, and should not change result.

SymverAliasMap on other hand, also depends on raw pointers values, which controlled by memory allocator, which are very inconsistent, especially with ASLR.

So could you please confirm that `SymverAliasMap` was the root-cause of your issue, and not `Symbols`?

Comment (llvmssh) [reply llvm/lib/Object/RecordStreamer.h:32]: Ok, thanks for your reply, Symbols may not need to change StringMap to MapVec. I'll check again if modifying only SymverAliasMap is OK.

Comment (llvmssh) [reply llvm/lib/Object/RecordStreamer.h:32]: I've made a confirmation that SymverAliasMap was the root-cause of your issue, and not Symbols.

**Review by teresajohnson — APPROVED** [review]
lgtm

File: llvm/lib/Object/RecordStreamer.h (hunk 1 above)
Comment (vitalybuka) [reply llvm/lib/Object/RecordStreamer.h]: Thank you!

**Review by vitalybuka — APPROVED** [review]

**Comment by vitalybuka** [comment]
I will land when buildkite is done.

**Comment by vitalybuka** [comment]
Thanks for improving consistency!

**Comment by llvmssh** [comment]
> @llvmssh Congratulations on having your first Pull Request (PR) merged into the LLVM Project!
> 
> Your changes will be combined with recent changes from other authors, then tested by our [build bots](https://lab.llvm.org/buildbot/). If there is a problem with a build, you may receive a report in an email or a comment on this PR.
> 
> Please check whether problems have been caused by your change specifically, as the builds can include changes from many authors. It is not uncommon for your change to be included in a build that fails due to someone else's changes, or infrastructure issues.
> 
> How to do this, and the rest of the post-merge process, is covered in detail [here](https://llvm.org/docs/MyFirstTypoFix.html#myfirsttypofix-issues-after-landing-your-pr).
> 
> If your change does cause a problem, it may be reverted, or you can revert it yourself. This is a normal part of [LLVM development](https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy). You can fix your changes and open a new PR to merge them again.
> 
> If you don't get any reports, no action is required from you. Your changes are working as expected, well done!

The Buildbot has detected a new failure on builder sanitizer-x86_64-linux-android while building llvm.

Full details are available at:
    https://lab.llvm.org/buildbot/#/builders/186/builds/4513

Worker for this Build: sanitizer-buildbot-android
Blamelist:
    Alan Phipps <[a-phipps@ti.com](mailto:a-phipps@ti.com)>,
    Justin Bogner <[mail@justinbogner.com](mailto:mail@justinbogner.com)>,
    llvmssh <[shaosenhao@huawei.com](mailto:shaosenhao@huawei.com)>

BUILD FAILED: failed 'python ../sanitizer_buildbot/sanitizers/zorg/buildbot/builders/sanitizers/buildbot_selector.py' (failure)

Step 2 (annotate) failure: 'python ../sanitizer_buildbot/sanitizers/zorg/buildbot/builders/sanitizers/buildbot_selector.py' (failure) ...
PASS: HWAddressSanitizer-aarch64 :: TestCases/hwasan_symbolize.cpp (527 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/Posix/ignore_free_hook.cpp (528 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/hwasan_symbolize_stack_uas.cpp (529 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/libc_thread_freeres.c (530 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/heap-buffer-overflow.c (531 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/longjmp-out-of-range.c (532 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/deep-recursion.c (533 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/malloc-test.c (534 of 1721)
PASS: HWAddressSanitizer-aarch64 :: TestCases/malloc-align.c (535 of 1721)
PASS: AddressSanitizer-aarch64-android :: TestCases/use-after-scope-types.cpp (536 of 1721)
FAIL: HWAddressSanitizer-aarch64 :: TestCases/hwasan_symbolize_stack_overflow.cpp (537 of 1721)
******************** TEST 'HWAddressSanitizer-aarch64 :: TestCases/hwasan_symbolize_stack_overflow.cpp' FAILED ******************** Exit Code: 1
I have a problem, I can't see anything in the link above, can't check if my modification caused the above error. @llvmbot @jh7370

**Comment by vitalybuka** [comment]
Probably a flake.

**Comment by llvmssh** [comment]
> Probably a flake.

Okay, thanks for your reply. Is it necessary to confirm this?

**Comment by vitalybuka** [comment]
> > Probably a flake.
> 
> Okay, thanks for your reply. Is it necessary to confirm this?

Multiple following builds have no this error.

**Comment by llvmssh** [comment]
> > > Probably a flake.
> > 
> > 
> > Okay, thanks for your reply. Is it necessary to confirm this?
> 
> Multiple following builds have no this error and the patch is still there. https://lab.llvm.org/buildbot/#/builders/186

Ok, thanks.

**Pull request merged**